### ISSUE TITLE
Add support for stable (PCI) addresses and introspection

### ISSUE THREAD

**Issue by fabiand** [kubevirt/kubevirt#1325]
> As a user I expect that my guest devices (and their location) only change if I modify the VM definition.

IOW: The guest ABI should have stable addresses.

This is a mandatory feature for VM identity, and avoiding that certain guests need reactivatio, i.e. Microsoft Windows

**Comment by ronf** [comment]
I need this capability as well, but for disk devices, as the OS I'm running in the virtual machine expects content to appear on specific PCI addresses. More specifically, it's a custom boot loader and it's having trouble finding the images to boot.

When I start up a VM with libvirt directly, it works when I specify something like:

    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2'/>
      <source file='/var/lib/libvirt/images/my_image.qcow2'/>
      <target dev='vda' bus='virtio'/>
      <address type='pci' domain='0x0000' bus='0x00' slot='0x05' function='0x0'/>
    </disk>

However, with kubevirt the PCI addresses are automatically selected and I end up with something like:

    <disk type='file' device='disk'>
      <driver name='qemu' type='qcow2'/>
      <source file='/var/run/libvirt/kubevirt-ephemeral-disk/registry-disk-data/default/sgva/disk_boot/disk-image.qcow2'/>
      <target dev='vda' bus='virtio'/>
      <address type='pci' domain='0x0000' bus='0x03' slot='0x00' function='0x0'/>
    </disk>

**Comment by fabiand** [comment]
That's interesting - we wondered about the disk case. So far we knew that the serials were the primary guest sided identifier. Interesting to see the use-case where the PCI address is relevant.

**Comment by ronf** [comment]
Yeah - I know it's a bit of an edge case (particularly for disks), but it'd be great if there was a way to specify the address information for any PCI device. I'm not sure yet if I'll run into a similar issue with the NICs, but I suspect I'll need it there too.

**Comment by fabiand** [comment]
@vladikr we also need pciAddresses for network devices, as they influence the interface name.

**Comment by vladikr** [comment]
> @vladikr we also need pciAddresses for network devices, as they influence the interface name.

@fabiand did you mean pciAddresses for disks? #1393 adds a support for setting pciAddresses for network devices.

**Comment by vladikr** [comment]
I'm working on allowing setting pciAddresses for disks.

**Comment by MarSik** [comment]
Just to resurrect the thread a bit. Commonly users do not want to specify the address beforehand, but would expect the sane behaviour after restart (no extra driver re-installation, same device ordering). Can we open the discussion about how to achieve the stable addressing the title of the issue mentions?

**Comment by rmohr** [comment]
@MarSik if we really have to, here some choices on how we can express it, which come to my mind:

 1. add it to the `status` of a `VMI` and the `VM`. One special thing to consider here, which we would have to solve, is to re-create a VMI from a `VM` with a pre-populated status (the privileged controller user would be able to pre-populate the VMI status basically with data from old runs). From the Spec perspective it would still be declarative and normal users would not be able to use it.
 2. Have some kind of a shadow CRD which would be considered by the VM and the VMI if present
 3. Add it to the `status` of the `VMI` and the `VM`. The `status` of the VM would be used to back-propagate the assigned addresses to the `VM` status, but this time we would also make sure that all the fields are also configurable on the spec of the `VMI`. After the first `VMI` start, the existing `VM` would take the backpropagated values from its own status and apply it to the `VMI` spec. In this case it would also be usable for external users.

@michalskrivanek there was always the idea floating around that only *maybe* we need to collect the addresses. Did we come to a final conclusion here that it is really necessary? On the community meeting we expressed the desire to see some issues (maybe you know some in bugzilla) which definitely required it.

/cc @vladikr FYI

**Comment by ronf** [comment]
For my use case, it was definitely important to be able to set the exact PCI addresses. The code running in the VM expected devices at specific locations and failed if it didn't find them where it was expecting them.

**Comment by rmohr** [comment]
@ronf I hope that you will be all set once #1484 from @vladikr  is merged. You can then set pci addresses for disks and interfaces. Your use-case would then not depend on the outcome of this discussion.

**Comment by ronf** [comment]
Thanks, @rmohr - I think you're right.
